### PR TITLE
Add request signer for messagebird webhook verification

### PR DESCRIPTION
**Pull request by klpx**
Completely ported from java library.
I think tests are enough for usage demonstration, aren't they?

### CODE REVIEW

File: MessageBird/Objects/Request.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+using System;
+using System.Text;
+
+namespace MessageBird.Objects
+{
+    public class Request
+    {
+        public string Timestamp { get; }
```
Comment (epels) [inline MessageBird/Objects/Request.cs]: I would prefer not making these properties part of the public API. Can we make this `internal`?

File: MessageBird/Objects/Request.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+using System;
+using System.Text;
+
+namespace MessageBird.Objects
+{
+    public class Request
```
Comment (epels) [inline MessageBird/Objects/Request.cs]: Perhaps add a comment here to clarify what the class is used for. Out of context, it is not clear from the code that it's used for (request) signing.

File: MessageBird/Objects/Request.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+         * Constructs a new request instance.
+         *
+         * @param timestamp Timestamp provided in the MessageBird-Request-Timestamp
+         *                  header.
+         * @param queryParameters Query parameters in abc=foo&def=ghi format.
+         * @param data Raw body of this request.
+         */
+        public Request(string timestamp, string queryParameters, byte[] data) {
+            if (string.IsNullOrEmpty(timestamp)) {
+                throw new ArgumentNullException(nameof(timestamp));
```
Comment (epels) [inline MessageBird/Objects/Request.cs]: As per the README, we need to support .NET Framework >=4.0 and .NET Standard 2.0 - I think we can't use `nameof`?

File: MessageBird/Objects/Request.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            QueryParameters = queryParameters;
+            Data = data;
+        }
+        
+        
+        public string SortedQueryParameters() {
+            var queryParams = QueryParameters.Split(QueryParametersDelimiter);
+            Array.Sort(queryParams);
+            var sortedParamsAccumulator = new StringBuilder();
+            for (int i = 0, paramsLength = queryParams.Length; i < paramsLength; i++) {
```
Comment (epels) [inline MessageBird/Objects/Request.cs]: `String.Join()`?

File: MessageBird/Objects/Request.cs (hunk 1 above)
Comment (klpx) [reply MessageBird/Objects/Request.cs]: Also, SortedQueryParameters()?

Comment (epels) [reply MessageBird/Objects/Request.cs]: Yes please!

File: MessageBird/Utilities/RequestSigner.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using MessageBird.Objects;
+
+namespace MessageBird.Utilities
```
[the remark this answers is not in the document]
Comment (klpx) [reply MessageBird/Utilities/RequestSigner.cs]: Sounds ok.

File: MessageBird/Objects/Request.cs (hunk 3 above)
Comment (klpx) [reply MessageBird/Objects/Request.cs]: Good point! Thanks!

File: MessageBird/Objects/Request.cs (hunk 4 above)
Comment (klpx) [reply MessageBird/Objects/Request.cs]: Thanks)

**Comment by klpx** [comment]
@epels  Made all proposals

File: MessageBird/Objects/Request.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                throw new ArgumentNullException("timestamp");
+            }
+
+            Timestamp = timestamp;
+            QueryParameters = queryParameters;
+            Data = data;
+        }
+        
+        
+        public string SortedQueryParameters() {
```
Comment (epels) [inline MessageBird/Objects/Request.cs]: Still public

File: MessageBird/Objects/Request.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Text;
+
+namespace MessageBird.Objects
+{
+    /**
+     * Structure used for request signature verification via MessageBird.Utilities.RequestSigner
```
Comment (epels) [inline MessageBird/Objects/Request.cs]: Should now be `MessageBird.RequestSigner`

File: MessageBird/Objects/Request.cs (hunk 6 above)
Comment (klpx) [reply MessageBird/Objects/Request.cs]: Thanks!

File: MessageBird/Objects/Request.cs (hunk 7 above)
Comment (klpx) [reply MessageBird/Objects/Request.cs]: Oops

**Pull request merged**